### ISSUE TITLE
Type.AllReferencedTypes includes types used by methods of the specified type. Is that what is desired?

### ISSUE THREAD

**Issue by gregveres** [NeVeSpl/NTypewriter#21]
Following your suggestion in issue #8, I switched over to use Type.AllReferencedTypes. 

Here is an example where too many types are returned. The import section is driven by the call to Types.AllRelatedTypes. As you can see, SessionBox and SeesionPlayer are not needed in this file. Ie, they are not directly related to BoxLeagueNextSession. They are related to BoxLeagueSession. 

```JS
import { NextSessionBox } from './NextSessionBox';
import { MemberInfo } from './MemberInfo';
import { SessionBox } from './SessionBox';
import { SessionPlayer } from './SessionPlayer';
import { BoxLeagueSession } from './BoxLeagueSession';


export interface BoxLeagueNextSession extends BoxLeagueSession {
  nextSession: NextSessionBox[],
  nextSessionDateRangeDescription: string,
  nextSessionStartDate: string,
  nextSessionEndDate: string,
  targetBoxSize: number,
  playersLeaving: MemberInfo[]
}
```

**Comment by NeVeSpl** [comment]
Could you provide fragments of c# code of BoxLeagueSession type where SessionBox and SeesionPlayer are used?

**Comment by NeVeSpl** [comment]
I am still not able to reproduce the bug, could you also show a fragment of the template that generates imports?

**Comment by gregveres** [comment]
Sure. And I just verified that I am still getting the output that I originally posted. (the template file has been through a lot of iterations in the last few days). 

My output is exactly the same as what I originally posted. 

Here is the template file:

```
{{- # Helper classes }}
{{- func ImportType(type)
	useType = Type.Unwrap(type)
	if (useType.ArrayType != null) 
	  useType = useType.ArrayType
	end
	if (type.IsEnum) 
	  path = "../Enums/"
	else
	  path = "./"
    end
	if ((useType.Attributes | Array.Filter @AttrIsExportToTypescript).size > 0)
		ret "import { " | String.Append useType.Name | String.Append " } from '" | String.Append path | String.Append useType.Name | String.Append "';\r"
	end
	ret null
	end
}}
{{- func AttrIsExportToTypescript(attr)
	ret attr.Name | String.Contains "ExportToTypescript"
	end
}}

{{- # output classes }}
{{- for class in data.Classes | Symbols.ThatHaveAttribute "ExportToTypescript" | Array.Concat (data.Classes | Symbols.ThatHaveAttribute "ExportToTypescriptWithKnockout")
		capture output }}

{{- for type in (class | Type.AllReferencedTypes)}}
	{{- ImportType type}}
{{-end}}

export interface {{class.Name}}{{if class.HasBaseClass}} extends {{class.BaseClass.Name; end}} {
  {{- for prop in class.Properties | Symbols.ThatArePublic }}
  {{ prop.Name }}: {{prop.Type | Type.ToTypeScriptType}}{{if !for.last}},{{end}}
  {{-end}}
}
{{- end}}
{{- Save output ("Interfaces\\" | String.Append class.BareName | String.Append ".ts")}}
{{- end}}
```

**Comment by gregveres** [comment]
I don't know what timezone you are in, but if you give up for the night, I can take a crack at creating a unit test case that covers it and seeing if I can reproduce it in the NTypewriter project. Just let me know when you stop working on it for the day.

**Comment by gregveres** [comment]
I think I have figured out where the extra references are coming from. 

When I pasted the classes, in the issue here, I did not include the methods. The two extra class references are used in return types of methods on BoxLeagueNextSession. 

I went through all 403 generated interface files and only 2 of them had extra imports. Both of the classes being exported have functions on them that reference the extra class as a return type.

Now that I know where the issue is coming from, I am wondering if types referenced in methods should be included. I don't really know the answer. I don't use it that way but someone else might use it that way. 

I then considered "should the function take an argument to specify the inclusion or exclusion of method types?". But since it was 2 files out of 403, I am not sure it matters to me. From a perfectionist point of view, I would like to not include them, but from a practical point of view, there are other things to get working first - like proper return types on controller actions. 

I will leave it to you. Now you know why those extra types are there.
